src/index/LocatedTriples.h

```diff
@@ -0,0 +1,174 @@
+// this to store all located triples with the same `blockIndex_`.
+//
+// NOTE: We could also overload `std::less` here, but the explicit specification
+// of the order makes it clearer.
+struct LocatedTripleCompare {
+  bool operator()(const LocatedTriple& x, const LocatedTriple& y) const {
+    return x.triple_ < y.triple_;
+  }
+};
+using LocatedTriples = std::set<LocatedTriple, LocatedTripleCompare>;
+
+std::ostream& operator<<(std::ostream& os, const LocatedTriples& lts);
+
+// Sorted sets of located triples, grouped by block. We use this to store all
+// located triples for a permutation.
+class LocatedTriplesPerBlock {
+ private:
+  // The total number of `LocatedTriple` objects stored (for all blocks).
+  size_t numTriples_ = 0;
+
+ public:
+  // For each block with a non-empty set of located triples, the located triples
+  // in that block.
+  //
+  // NOTE: This is currently not private because we want access to
+  // `map_.size()`, `map_.clear()`, `map_.contains(...)`, and `map_.at(...)`.
+  // We could also make `LocatedTriplesPerBlock` a subclass of `HashMap<size_t,
+  // LocatedTriples>`, but not sure whether that is good style.
+  ad_utility::HashMap<size_t, LocatedTriples> map_;
+
+  // Get upper limits for the number of located triples for the given block. The
+  // return value is a pair of numbers: first, the number of existing triples
+  // ("to be deleted") and second, the number of new triples ("to be inserted").
+  // The numbers are only upper limits because there may be triples that have no
+  // effect (like adding an already existing triple and deleting a non-existent
+  // triple).
+  std::pair<size_t, size_t> numTriples(size_t blockIndex) const;
+
+  // Merge located triples for `blockIndex_` with the given index `block` and
+  // write to `result`, starting from position `offsetInResult`. Return the
+  // number of rows written to `result`.
+  //
+  // PRECONDITIONS:
+  //
+  // 1. The set of located triples for `blockIndex_` must be non-empty.
+  // Otherwise, there is no need for merging and this method shouldn't be
+  // called for efficiency reasons.
+  //
+  // 2. It is the responsibility of the caller that there is enough space for
+  // the result of the merge in `result` starting from `offsetInResult`.
+  size_t mergeTriples(size_t blockIndex, IdTable block, IdTable& result,
+                      size_t offsetInResult, size_t numIndexColumns) const;
+
+  // Add `locatedTriples` to the `LocatedTriplesPerBlock`.
+  // Return handles to where they were added (`LocatedTriples` is a sorted set,
+  // see above). We need the handles so that we can easily remove the
+  // `locatedTriples` from the set again in case we need to.
+  //
+  // PRECONDITIONS:
+  //
+  // 1. The `locatedTriples` must not already exist in
+  // `LocatedTriplesPerBlock`.
+  std::vector<LocatedTriples::iterator> add(
+      const std::vector<LocatedTriple>& locatedTriples);
+
+  void erase(size_t blockIndex, LocatedTriples::iterator iter);
```

Comment:
This code is from the original draft PR, so I can only guess the reasons. It is more efficient to delete items using the iterator directly. `DeltaTriples` has the triples in a hash_map anyways. They are retrieved to ensure that Triples are not added/deleted twice or both added and deleted. Storing and retrieving the iterators in `DeltaTriples` can be done at no additional cost.

Reply:
If the `DeltaTriples` stored the `LocatedTriple` instead of the iterator, we could get rid of one parameter. And the lookup here would be per Block while it is globally in `DeltaTriples`. So the performance impact is negligible. The only thing that's left is that the `LocatedTriple` is larger. TLDR: lets get rid of the iterators.